### PR TITLE
Registration Gating (Proof-of-Work, Terms of Service)

### PR DESCRIPTION
**Pull request by finn-tbd**
This PR adds optional (disabled by default) authorization to all endpoints. New DIDs wishing to use an instance of dwn-server with registration requirements must complete a proof-of-work challenge, agree to the server's terms of service, or both. Requests for DIDs that have not completed the requirements will be rejected.

POW authorization can be enabled by setting `DWN_REGISTRATION_POW=true`. Authorized DIDs are tracked in a database, configured by setting `DWN_STORAGE_REGISTRATION` or `DWN_STORAGE`. Note that only SQL-based stores are supported.

### CODE REVIEW

File: tests/scenarios/registration.spec.ts [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+  afterEach(async function () {
+    server.close();
+    server.closeAllConnections();
+  });
+
+  after(function () {
+    clock.restore();
+  });
+
+  it('should facilitate tenant registration with terms-of-service and proof-or-work turned on', async () => {
```
Comment (thehenrytsai) [inline tests/scenarios/registration.spec.ts:86]: @shamilovtim, this shows how tenant registration is done!

**Comment by LiranCohen** [comment]
I've given it a first pass (minus the tests) and it looks good so far when analyzing the individual pieces.  Want to spend a little more time today re-reviewing the PoW calculation pieces and giving it a last pass including tests.

Just a few small nits and questions above.

File: src/http-api.ts [hunk 2]
```diff
@@ -181,13 +171,68 @@ export class HttpApi {
         return res.json(jsonRpcResponse);
       }
     });
+
+    this.#setupRegistrationRoutes();
+
+    this.#api.get('/info.json', (req, res) => {
```
[the remark this answers is not in the document]
Comment (thehenrytsai) [reply src/http-api.ts:177]: Actually @finn-tbd added this, though, imma change it to just `/info`!

File: src/registration/proof-of-work-manager.ts [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+  private proofOfWorkOfLastMinute: Map<string, number> = new Map(); // proofOfWorkId -> timestamp of proof-of-work
+
+  private currentMaximumHashValueAsBigInt: bigint;
+  private initialMaximumHashValueAsBigInt: bigint;
+  private desiredSolveCountPerMinute: number;
+
+  /**
+   * How often the challenge nonce is refreshed.
+   */
+  public challengeRefreshFrequencyInSeconds: number;
```
[the remark this answers is not in the document]
Comment (thehenrytsai) [reply src/registration/proof-of-work-manager.ts:23]: The thinking is that `challengeRefreshFrequencyInSeconds` and `difficultyReevaluationFrequencyInSeconds` can be externally read or modified.

They are already access in tests.

File: src/registration/proof-of-work-manager.ts [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+
+  /**
+   * How often the difficulty is reevaluated.
+   */
+  public difficultyReevaluationFrequencyInSeconds: number;
```
[the remark this answers is not in the document]
Comment (thehenrytsai) [reply src/registration/proof-of-work-manager.ts:28]: Same as the above comment.

File: src/registration/proof-of-work-manager.ts [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+
+    const latestSolveCountPerMinute = this.proofOfWorkOfLastMinute.size;
+
+    // NOTE: bigint arithmetic does NOT work with decimals, so we work with "full numbers" by multiplying by a scale factor.
+    const scaleFactor = 1_000_000;
+    const difficultyEvaluationsPerMinute = 60000 / (this.difficultyReevaluationFrequencyInSeconds * 1000); // assumed to be >= 1;
+
+    // NOTE: easier difficulty is represented by a larger max allowed hash value
+    //       and harder difficulty is represented by a smaller max allowed hash value.
+    const currentSolveRateInFractionOfDesiredSolveRate = latestSolveCountPerMinute / this.desiredSolveCountPerMinute;
```
[the remark this answers is not in the document]
Comment (thehenrytsai) [reply src/registration/proof-of-work-manager.ts:208]: You are right, can be inside, originally the difficulty decreasing code (in `else`) also used it, but that's no longer the case!

Done.

File: src/registration/proof-of-work-manager.ts [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+    if (latestSolveCountPerMinute > this.desiredSolveCountPerMinute) {
+      this.hashValueIncrementPerEvaluation = undefined;
+
+      // if solve rate is higher than desired, make difficulty harder by making the max allowed hash value smaller
+      
+      // set higher to make difficulty increase faster.
+      // This should also be relative to how often the difficulty is reevaluated if the reevaluation frequency is adjustable.
+      const increaseMultiplier = 1;
```
[the remark this answers is not in the document]
Comment (thehenrytsai) [reply src/registration/proof-of-work-manager.ts]: The intend here was a TODO to expose this as an adjustable parameter. Anything above 1 will have an exponential effect.

I have gone ahead to expose it in the `ProofOfWorkManager` to show the intent, and created issue #100 to track the exposing of this as a server setting.

File: src/registration/proof-of-work-manager.ts [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+
+      const newMaximumHashValueAsBigIntPriorToMultiplierAdjustment
+        = (this.currentMaximumHashValueAsBigInt * BigInt(scaleFactor)) / 
+          (BigInt(Math.floor(currentSolveRateInFractionOfDesiredSolveRate * increaseMultiplier * scaleFactor)));
+
+      // set higher to make difficulty increase faster.
+      // This should also be relative to how often the difficulty is reevaluated if the reevaluation frequency is adjustable.
```
[the remark this answers is not in the document]
Comment (thehenrytsai) [reply src/registration/proof-of-work-manager.ts]: Moved to a better location and gave better description.

File: src/registration/proof-of-work-manager.ts [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+      
+      // set higher to make difficulty increase faster.
+      // This should also be relative to how often the difficulty is reevaluated if the reevaluation frequency is adjustable.
+      const increaseMultiplier = 1;
+
+      const newMaximumHashValueAsBigIntPriorToMultiplierAdjustment
+        = (this.currentMaximumHashValueAsBigInt * BigInt(scaleFactor)) / 
+          (BigInt(Math.floor(currentSolveRateInFractionOfDesiredSolveRate * increaseMultiplier * scaleFactor)));
+
+      // set higher to make difficulty increase faster.
```
[the remark this answers is not in the document]
Comment (thehenrytsai) [reply src/registration/proof-of-work-manager.ts]: Removed.

File: src/registration/proof-of-work-manager.ts [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+    this.challengeNonces.currentChallengeNonce = ProofOfWork.generateNonce();
+  }
+
+  /**
+   * Refreshes the difficulty by changing the max hash value.
+   * The higher the number, the easier. Scale 1 (hardest) to 2^256 (easiest), represented in HEX.
+   * 
+   * If solve rate rate is higher than expected, the difficulty will increase rapidly.
+   * If solve rate is lower than expected, the difficulty will decrease gradually.
+   * The difficulty will never be lower than the initial difficulty.
```
[the remark this answers is not in the document]
Comment (thehenrytsai) [reply src/registration/proof-of-work-manager.ts:203]: You understand correctly, though the method description is describing the intent which is talking about "difficulty" in general. The actually implementation detail uses the hash value (ie. leading zeros), which happens to be inverse of "difficulty".

[the remark this answers is not in the document]
Comment (LiranCohen) [reply src/registration/proof-of-work-manager.ts:208]: Doh! Yeah sorry that makes sense! Was in the weeds thinking about the hash value, but yeah more difficult is higher although it's lower hash.

File: tests/http-api.spec.ts [hunk 10]
```diff
@@ -254,11 +284,11 @@ describe('http api', function () {
       expect(body.error).to.not.exist;
 
       const { reply } = body.result;
+      console.log(reply);
```
Comment (LiranCohen) [inline tests/http-api.spec.ts:287]: Are we logging this here on purpose during testing? If necessary, should we add some label?

File: tests/registration/proof-of-work-manager.spec.ts [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+    const initialMaximumAllowedHashValue = 'FFFFFFFF';
+    const proofOfWorkManager = await ProofOfWorkManager.create({
+      autoStart: true,
+      desiredSolveCountPerMinute,
+      initialMaximumAllowedHashValue,
+    });
+
+    // stub that throws half the time
+    const stub = (): void => {
+      // Generate a random number between 0 and 1
```
Comment (LiranCohen) [inline tests/registration/proof-of-work-manager.spec.ts:38]: This is a really interesting part of the test. Do you think testing the potential failures and their effect is suitable for a different test?

Or perhaps just providing some context in comments to how the system would respond to these scenarios?

File: tests/registration/proof-of-work-manager.spec.ts [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+      // If the random number is less than 0.5, throw an error
+      if (random < 0.5) {
+        throw new Error('Random error');
+      }
+    };
+
+    const challengeNonceRefreshSpy = sinon.stub(proofOfWorkManager, 'refreshChallengeNonce').callsFake(stub);
+    const maximumAllowedHashValueRefreshSpy = sinon.stub(proofOfWorkManager, 'refreshMaximumAllowedHashValue').callsFake(stub);
+
+    // Simulated 1 hour has passed, so all proof-of-work entries should be removed.
```
Comment (LiranCohen) [inline tests/registration/proof-of-work-manager.spec.ts:50]: Not sure I follow if this comment is relevant to this test. I also am not sure if I understand what the comment is describing.

Is this referring to the map which holds the proof of work in the last minute? If so the hour comment makes it a bit confusing, and not sure that it makes sense to mention here.

File: tests/registration/proof-of-work-manager.spec.ts [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+    // Load up desiredSolveRatePerMinute number of proof-of-work entries, so all future new entries will increase the complexity.
+    for (let i = 0; i < desiredSolveCountPerMinute; i++) {
+      await proofOfWorkManager.recordProofOfWork(uuidv4());
+    }
+
+    let baselineMaximumAllowedHashValue = proofOfWorkManager.currentMaximumAllowedHashValue;
+    let lastMaximumAllowedHashValue = BigInt('0x' + initialMaximumAllowedHashValue);
+    const lastSolveCountPerMinute = 0;
+    for (let i = 0; i < 100; i++) {
+      // Simulating 1 proof-of-work per second which for 100 seconds.
```
Comment (LiranCohen) [inline tests/registration/proof-of-work-manager.spec.ts:81]:
```suggestion
      // Simulating 1 proof-of-work per second for 100 seconds.
```

File: tests/registration/proof-of-work-manager.spec.ts [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+      desiredSolveCountPerMinute,
+      initialMaximumAllowedHashValue,
+    });
+
+    // Load up desiredSolveRatePerMinute number of proof-of-work entries, so all future new entries will increase the complexity.
+    for (let i = 0; i < desiredSolveCountPerMinute; i++) {
+      await proofOfWorkManager.recordProofOfWork(uuidv4());
+    }
+
+    // Simulating 1 proof-of-work per second which for 100 seconds to increase proof-of-work difficulty.
```
Comment (LiranCohen) [inline tests/registration/proof-of-work-manager.spec.ts:125]:
```suggestion
    // Simulating 1 proof-of-work per second for 100 seconds to increase proof-of-work difficulty.
```

File: tests/scenarios/registration.spec.ts [hunk 15]
```diff
@@ -0,0 +1,455 @@
...
+    const write2Response = await fetch(dwnMessageEndpoint, {
+      method: 'POST',
+      headers: {
+        'dwn-request': JSON.stringify(write2.jsonRpcRequest),
+      },
+      body: new Blob([write2.dataBytes]),
+    });
+    const write2ResponseBody = await write2Response.json() as JsonRpcResponse;
+    expect(write2Response.status).to.equal(200);
+    expect(write2ResponseBody.result.reply.status.code).to.equal(401);
```
Comment (LiranCohen) [inline tests/scenarios/registration.spec.ts:440]: Sanity: consider adding to this test where Alice re-registers and can now write the same DWN message.

**Review by LiranCohen — CHANGES_REQUESTED** [review]
Just a few small nits in the tests.

This looks good, and the added TODOs will make it even better when necessary. 🚀 🐐

File: tests/http-api.spec.ts (hunk 10 above)
Comment (thehenrytsai) [reply tests/http-api.spec.ts:287]: No, must've forgotten to remove after some sanity testing.

Removed.

File: tests/registration/proof-of-work-manager.spec.ts (hunk 11 above)
Comment (thehenrytsai) [reply tests/registration/proof-of-work-manager.spec.ts:38]: I've update the test title to better describe the intent of the test. Let me know if you think I should add or update anything else.

File: tests/registration/proof-of-work-manager.spec.ts (hunk 12 above)
Comment (thehenrytsai) [reply tests/registration/proof-of-work-manager.spec.ts:50]: You are right! Completely useless and irrelevant, it's originally a C&P from another test. Removed!

File: tests/registration/proof-of-work-manager.spec.ts (hunk 13 above)
Comment (thehenrytsai) [reply tests/registration/proof-of-work-manager.spec.ts:81]: Thanks! Fixed!

File: tests/scenarios/registration.spec.ts (hunk 15 above)
Comment (thehenrytsai) [reply tests/scenarios/registration.spec.ts:440]: Good call! Adding now.

**Review by LiranCohen — APPROVED** [review]
LGTM! 🚀

**Pull request merged**